linters-settings.thelper.test.name not working

### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc).
- [X] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

### Description of the problem

It's impossible to disable thelper's "t_name" check using `.golangci.yml`.

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.46.0 built from 7c1d8b75 on 2022-05-09T11:16:24Z
```

</details>

### Configuration file

<details>

```console
$ cat .golangci.yml
linters-settings:
  thelper:
    test:
      name: false

linters:
  disable-all: true
  enable:
    - thelper
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.18.2 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/powerman/.cache/go-build"
GOENV="/home/powerman/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/powerman/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/powerman/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/lib/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.18.2"
GCCGO="gccgo"
GOAMD64="v3"
AR="ar"
CC="x86_64-pc-linux-gnu-gcc"
CXX="x86_64-pc-linux-gnu-g++"
CGO_ENABLED="1"
GOMOD="/home/powerman/tmp/go/bug-golangci-lint-thelper/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3888433916=/tmp/go-build -gno-record-gcc-switches"
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
bug_test.go:5:6: parameter *testing.T should have name t (thelper)                               
func helper(tt *testing.T) {
     ^
```

</details>

### Code example or link to a public repository

<details>

```go
package bug_test

import "testing"

func helper(tt *testing.T) {
	tt.Helper()
}

func Test(tt *testing.T) {
	tt.Parallel()
	helper(tt)
}
```

</details>

```
$ thelper -checks=b_begin,b_first,b_name,t_begin,t_first,tb_begin,tb_first,tb_name .
$ thelper -checks=b_begin,b_first,b_name,t_begin,t_first,t_name,tb_begin,tb_first,tb_name .                                                                         
/home/powerman/tmp/go/bug-golangci-lint-thelper/bug_test.go:5:6: parameter *testing.T should have name t
```

--- a/pkg/config/linters_settings.go
+++ b/pkg/config/linters_settings.go
@@ -550,26 +550,16 @@ type TestpackageSettings struct {
 }
 
 type ThelperSettings struct {
-	Test struct {
-		First bool `mapstructure:"first"`
-		Name  bool `mapstructure:"name"`
-		Begin bool `mapstructure:"begin"`
-	} `mapstructure:"test"`
-	Fuzz struct {
-		First bool `mapstructure:"first"`
-		Name  bool `mapstructure:"name"`
-		Begin bool `mapstructure:"begin"`
-	} `mapstructure:"fuzz"`
-	Benchmark struct {
-		First bool `mapstructure:"first"`
-		Name  bool `mapstructure:"name"`
-		Begin bool `mapstructure:"begin"`
-	} `mapstructure:"benchmark"`
-	TB struct {
-		First bool `mapstructure:"first"`
-		Name  bool `mapstructure:"name"`
-		Begin bool `mapstructure:"begin"`
-	} `mapstructure:"tb"`
+	Test      ThelperOptions `mapstructure:"test"`
+	Fuzz      ThelperOptions `mapstructure:"fuzz"`
+	Benchmark ThelperOptions `mapstructure:"benchmark"`
+	TB        ThelperOptions `mapstructure:"tb"`
+}
+
+type ThelperOptions struct {
+	First *bool `mapstructure:"first"`
+	Name  *bool `mapstructure:"name"`
+	Begin *bool `mapstructure:"begin"`
 }
 
 type TenvSettings struct {

--- a/pkg/golinters/thelper.go
+++ b/pkg/golinters/thelper.go
@@ -13,53 +13,44 @@ import (
 func NewThelper(cfg *config.ThelperSettings) *goanalysis.Linter {
 	a := analyzer.NewAnalyzer()
 
-	cfgMap := map[string]map[string]interface{}{}
+	opts := map[string]struct{}{
+		"t_name":  {},
+		"t_begin": {},
+		"t_first": {},
+
+		"f_name":  {},
+		"f_begin": {},
+		"f_first": {},
+
+		"b_name":  {},
+		"b_begin": {},
+		"b_first": {},
+
+		"tb_name":  {},
+		"tb_begin": {},
+		"tb_first": {},
+	}
+
 	if cfg != nil {
-		var opts []string
+		applyTHelperOptions(cfg.Test, "t_", opts)
+		applyTHelperOptions(cfg.Fuzz, "f_", opts)
+		applyTHelperOptions(cfg.Benchmark, "b_", opts)
+		applyTHelperOptions(cfg.TB, "tb_", opts)
+	}
 
-		if cfg.Test.Name {
-			opts = append(opts, "t_name")
-		}
-		if cfg.Test.Begin {
-			opts = append(opts, "t_begin")
-		}
-		if cfg.Test.First {
-			opts = append(opts, "t_first")
-		}
+	if len(opts) == 0 {
+		linterLogger.Fatalf("thelper: at least one option must be enabled")
+	}
 
-		if cfg.Fuzz.Name {
-			opts = append(opts, "f_name")
-		}
-		if cfg.Fuzz.Begin {
-			opts = append(opts, "f_begin")
-		}
-		if cfg.Fuzz.First {
-			opts = append(opts, "f_first")
-		}
+	var args []string
+	for k := range opts {
+		args = append(args, k)
+	}
 
-		if cfg.Benchmark.Name {
-			opts = append(opts, "b_name")
-		}
-		if cfg.Benchmark.Begin {
-			opts = append(opts, "b_begin")
-		}
-		if cfg.Benchmark.First {
-			opts = append(opts, "b_first")
-		}
-
-		if cfg.TB.Name {
-			opts = append(opts, "tb_name")
-		}
-		if cfg.TB.Begin {
-			opts = append(opts, "tb_begin")
-		}
-		if cfg.TB.First {
-			opts = append(opts, "tb_first")
-		}
-
-		cfgMap[a.Name] = map[string]interface{}{
-			"checks": strings.Join(opts, ","),
-		}
+	cfgMap := map[string]map[string]interface{}{
+		a.Name: {
+			"checks": strings.Join(args, ","),
+		},
 	}
 
 	return goanalysis.NewLinter(
@@ -68,4 +59,24 @@ func NewThelper(cfg *config.ThelperSettings) *goanalysis.Linter {
 		[]*analysis.Analyzer{a},
 		cfgMap,
 	).WithLoadMode(goanalysis.LoadModeTypesInfo)
+}
+
+func applyTHelperOptions(o config.ThelperOptions, prefix string, opts map[string]struct{}) {
+	if o.Name != nil {
+		if !*o.Name {
+			delete(opts, prefix+"name")
+		}
+	}
+
+	if o.Begin != nil {
+		if !*o.Begin {
+			delete(opts, prefix+"begin")
+		}
+	}
+
+	if o.First != nil {
+		if !*o.First {
+			delete(opts, prefix+"first")
+		}
+	}
 }

--- a/test/testdata/configs/thelper.yml
+++ b/test/testdata/configs/thelper.yml
@@ -1,0 +1,5 @@
+linters-settings:
+  thelper:
+    test:
+      name: false
+      begin: true

--- a/test/testdata/thelper_with_options.go
+++ b/test/testdata/thelper_with_options.go
@@ -1,0 +1,18 @@
+// args: -Ethelper
+// config_path: testdata/configs/thelper.yml
+package testdata
+
+import "testing"
+
+func thelperWithHelperAfterAssignmentWO(t *testing.T) { // ERROR "test helper function should start from t.Helper()"
+	_ = 0
+	t.Helper()
+}
+
+func thelperWithNotFirstWO(s string, t *testing.T, i int) { // ERROR `parameter \*testing.T should be the first`
+	t.Helper()
+}
+
+func thelperWithIncorrectNameWO(o *testing.T) {
+	o.Helper()
+}

